packages/@react-stately/table/src/TableUtils.ts

```diff
@@ -133,42 +65,215 @@ export interface IColumn {
+ * @param getDefaultMinWidth - A function that returns the default min width of a column by its index.
+ */
 export function calculateColumnSizes(availableWidth: number, columns: IColumn[], changedColumns: Map<Key, ColumnSize>, getDefaultWidth, getDefaultMinWidth) {
-  let remainingSpace = availableWidth;
-  let {staticColumns, dynamicColumns} = columns.reduce((acc, column, index) => {
+  let flexItems = columns.map((column, index) => {
     let width = changedColumns.get(column.key) != null ? changedColumns.get(column.key) : column.width ?? column.defaultWidth ?? getDefaultWidth?.(index) ?? '1fr';
-    let minWidth = column.minWidth ?? getDefaultMinWidth?.(index);
-    column.minWidth = minWidth;
-
+    let frozen = false;
+    let baseSize = 0;
+    let flex = 0;
     if (isStatic(width)) {
-      let w = parseStaticWidth(width, availableWidth);
-      w = Math.max(
-        getMinWidth(column.minWidth, availableWidth),
-        Math.min(Math.floor(w), getMaxWidth(column.maxWidth, availableWidth)));
-      acc.staticColumns.push({index, column, width: w} as IndexedColumn);
-      acc.dynamicColumns.push(null);
-      remainingSpace -= w;
+      baseSize = parseStaticWidth(width, availableWidth);
+      frozen = true;
     } else {
-      let w = parseFractionalUnit(width);
-      acc.staticColumns.push(null);
-      acc.dynamicColumns.push({index, column, width: w} as IndexedColumn);
+      flex = parseFractionalUnit(width);
+      if (flex <= 0) {
+        frozen = true;
+      }
     }
-    return acc;
-  }, {staticColumns: [] as IndexedColumn[], dynamicColumns: [] as IndexedColumn[]});
-  let newColWidths = getDynamicColumnWidths(dynamicColumns, remainingSpace, availableWidth);
 
-  return staticColumns.map((col, i) => {
-    if (col) {
-      return col.width;
+    let min = getMinWidth(column.minWidth ?? getDefaultMinWidth?.(index) ?? 0, availableWidth);
+    let max = getMaxWidth(column.maxWidth, availableWidth);
+    let hypotheticalMainSize = Math.max(min, Math.min(baseSize, max));
+    return {
+      frozen,
+      baseSize,
+      hypotheticalMainSize,
+      min,
+      max,
+      flex,
+      targetMainSize: null,
+      violation: ''
+    };
+  });
+  // 9.7.1
+  // We don't make use of flex basis, it's always 0, so we are always in 'grow' mode.
+  // 9.7.2
+  flexItems.forEach(item => {
+    if (item.frozen) {
+      item.targetMainSize = item.hypotheticalMainSize;
+      return;
+    }
+    if (item.baseSize > item.hypotheticalMainSize) {
+      item.frozen = true;
+      item.targetMainSize = item.hypotheticalMainSize;
     }
-    return newColWidths[i].width;
   });
+
+  // 9.7.3
+  let initialFreeSpace = availableWidth - flexItems.reduce((acc, item) => {
+    if (item.frozen) {
+      return acc + item.targetMainSize;
+    }
+    return acc + item.baseSize;
+  }, 0);
+
+  let remainingFreeSpace = initialFreeSpace;
+  // 9.7.4
+  // 9.7.4.a
+  while (flexItems.some(item => !item.frozen)) {
```

Comment:
Maybe we could keep track of a boolean `hasNonFrozenItems` in the above loop, and update it in the last loop when marking items frozen or not? That would avoid another nested loop.